src/compiler/transformers/ts.ts

```diff
@@ -652,6 +655,9 @@ namespace ts {
                 startOnNewLine(varStatement);
                 statements = [varStatement];
             }
+            if (hoistedNameAssignments.length) {
+                statements.unshift(createVariableStatement(/*modifiers*/ undefined, map(hoistedNameAssignments, a => createVariableDeclaration(a.left as Identifier, /*type*/ undefined, a.right))));

```

Comment:
Seems a bit wasteful to create a `BinaryExpression` just to drop it. In _destructuring.ts_ we have to do something similar, so we pass a callback that creates the correct assignment in each context.